gcd/src/main/java/org/spine3/server/storage/datastore/DsEventStorage.java

```diff
@@ -20,100 +20,295 @@
+            if (entityResult == null) {
+                return EventStorageRecord.getDefaultInstance();
+            }
+            final EventStorageRecord message = entityToMessage(entityResult, RECORD_TYPE_URL);
+            return message;
+        }
+    };
+
+    private static final Function<Any, String> ID_TRANSFORMER = new Function<Any, String>() {
+        @Nullable
+        @Override
+        public String apply(@Nullable Any input) {
+            if (input == null) {
+                return null;
+            }
+
+            return Identifiers.idToString(input);
+        }
+    };
+
+    /* package */ static DsEventStorage newInstance(DatastoreWrapper datastore, boolean multitenant) {
+        return new DsEventStorage(datastore, multitenant);
     }
 
-    private DsEventStorage(DatastoreWrapper datastore) {
+    private DsEventStorage(DatastoreWrapper datastore, boolean multitenant) {
+        super(multitenant);
         this.datastore = datastore;
     }
 
+    private static Predicate<EventStorageRecord> eventPredicate(
+            @SuppressWarnings("TypeMayBeWeakened") EventStreamQuery query) {
+        return new EventPredicate(query.getFilterList());
+    }
+
     @Override
     public Iterator<Event> iterator(EventStreamQuery eventStreamQuery) {
+        final Query query = toTimestampQuery(eventStreamQuery);
 
-        final Query.Builder query = DatastoreQueries
-                .makeQuery(ASCENDING, KIND, eventStreamQuery);
-        final Iterator<EntityResult> iterator = datastore.runQueryForIterator(query.build());
-        final Iterator<Event> transformedIterator = Iterators.transform(iterator, ENTITY_TO_EVENT);
+        final Collection<Entity> entities = datastore.read(query);
+        // Transform and filter order does not matter since both operations are performed lazily
+        Collection<EventStorageRecord> events = Collections2.transform(entities, ENTITY_TO_EVENT_RECORD);
+        events = Collections2.filter(events, eventPredicate(eventStreamQuery));
 
-        return transformedIterator;
+        final Iterator<EventStorageRecord> iterator = events.iterator();
+        return toEventIterator(iterator);
+    }
+
+    @SuppressWarnings({"MethodWithMoreThanThreeNegations", "ValueOfIncrementOrDecrementUsed", "DuplicateStringLiteralInspection"})
+    private static Query toTimestampQuery(EventStreamQueryOrBuilder query) {
+        final long lower = Timestamps.convertToNanos(query.getAfter());
+        final long upper = query.hasBefore()
+                           ? Timestamps.convertToNanos(query.getBefore())
+                           : Long.MAX_VALUE;
+        final PropertyFilter greaterThen = PropertyFilter.gt(TIMESTAMP_NANOS_PROPERTY_NAME, lower);
+        final PropertyFilter lessThen = PropertyFilter.lt(TIMESTAMP_NANOS_PROPERTY_NAME, upper);
+        final CompositeFilter filter = CompositeFilter.and(greaterThen, lessThen);
+        return Query.newEntityQueryBuilder()
```

Comment:
Please have a `result` value before return.

Usage of this pattern through the code will make the debugging much easier.